File: app/src/main/java/io/github/zwieback/familyfinance/business/chart/fragment/PieChartOfExpensesFragment.java
package io.github.zwieback.familyfinance.business.chart.fragment;

import androidx.fragment.app.DialogFragment;

import io.github.zwieback.familyfinance.R;
import io.github.zwieback.familyfinance.business.chart.dialog.PieChartDisplayDialog;
import io.github.zwieback.familyfinance.business.operation.dialog.ExpenseOperationFilterDialog;
import io.github.zwieback.familyfinance.business.operation.filter.ExpenseOperationFilter;
import io.github.zwieback.familyfinance.business.operation.query.ExpenseOperationQueryBuilder;
import io.github.zwieback.familyfinance.core.model.OperationView;
import io.requery.query.Result;

public class PieChartOfExpensesFragment extends PieChartFragment<ExpenseOperationFilter> {

    @Override
    protected String getFilterName() {
        return ExpenseOperationFilter.EXPENSE_OPERATION_FILTER;
    }

    @Override
    protected ExpenseOperationFilter createDefaultFilter() {
        return new ExpenseOperationFilter();
    }

    @Override
    protected Result<OperationView> buildOperations() {
        return ExpenseOperationQueryBuilder.Companion.create(data)
                .withStartDate(filter.getStartDate())
                .withEndDate(filter.getEndDate())
                .withStartValue(filter.getStartValue())
                .withEndValue(filter.getEndValue())
                .withOwnerId(filter.getOwnerId())
                .withCurrencyId(filter.getCurrencyId())
                .withArticleId(filter.getArticleId())
                .withAccountId(filter.getAccountId())
                .build();
    }

    @Override
    protected int getDataSetLabel() {
        return R.string.data_set_expenses;
    }

    @Override
    public void showFilterDialog() {
        DialogFragment dialog = ExpenseOperationFilterDialog.Companion.newInstance(filter,
                R.string.pie_chart_of_expenses_filter_title);
        dialog.show(getChildFragmentManager(), "ExpenseOperationFilterDialog");
    }

    @Override
    public void showDisplayDialog() {
        DialogFragment dialog = PieChartDisplayDialog.newInstance(display,
                R.string.pie_chart_of_expenses_display_title);
        dialog.show(getChildFragmentManager(), "PieChartDisplayDialog");
    }
}

File: app/src/main/java/io/github/zwieback/familyfinance/business/chart/fragment/PieChartOfExpensesFragment.kt
package io.github.zwieback.familyfinance.business.chart.fragment

import io.github.zwieback.familyfinance.R
import io.github.zwieback.familyfinance.business.chart.dialog.PieChartDisplayDialog
import io.github.zwieback.familyfinance.business.operation.dialog.ExpenseOperationFilterDialog
import io.github.zwieback.familyfinance.business.operation.filter.ExpenseOperationFilter
import io.github.zwieback.familyfinance.business.operation.query.ExpenseOperationQueryBuilder
import io.github.zwieback.familyfinance.core.model.OperationView
import io.github.zwieback.familyfinance.extension.operation.filter.applyPreferences
import io.requery.query.Result

class PieChartOfExpensesFragment : PieChartFragment<ExpenseOperationFilter>() {

    override val filterName: String
        get() = ExpenseOperationFilter.EXPENSE_OPERATION_FILTER

    override val dataSetLabel: Int
        get() = R.string.data_set_expenses

    override fun createDefaultFilter() = ExpenseOperationFilter().apply {
        applyPreferences(this@PieChartOfExpensesFragment.requireContext())
    }

    override fun buildOperations(): Result<OperationView> {
        return ExpenseOperationQueryBuilder.create(data)
            .withStartDate(filter.startDate)
            .withEndDate(filter.endDate)
            .withStartValue(filter.startValue)
            .withEndValue(filter.endValue)
            .withOwnerId(filter.takeOwnerId())
            .withCurrencyId(filter.takeCurrencyId())
            .withArticleId(filter.takeArticleId())
            .withAccountId(filter.takeAccountId())
            .build()
    }

    override fun showFilterDialog() {
        ExpenseOperationFilterDialog
            .newInstance(filter, R.string.pie_chart_of_expenses_filter_title)
            .show(childFragmentManager, "ExpenseOperationFilterDialog")
    }

    override fun showDisplayDialog() {
        PieChartDisplayDialog
            .newInstance(display, R.string.pie_chart_of_expenses_display_title)
            .show(childFragmentManager, "PieChartDisplayDialog")
    }
}
